src/event_manager.rs

```diff
@@ -0,0 +1,493 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::fmt::Formatter;
+use std::io;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::sync::{Arc, Mutex};
+
+use crate::epoll::{self, Epoll, EpollEvent};
+
+pub type Result<T> = std::result::Result<T, Error>;
+pub type Pollable = RawFd;
+
+/// Errors associated with epoll events handling.
+pub enum Error {
+    /// Cannot create epoll fd.
+    EpollCreate(io::Error),
+    /// Polling I/O error.
+    Poll(io::Error),
+    /// The specified pollable already registered.
+    AlreadyExists(Pollable),
+    /// The specified pollable is not registered.
+    NotFound(Pollable),
+}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+
+        match self {
+            EpollCreate(err) => write!(f, "Unable to create epoll fd: {}", err),
+            Poll(err) => write!(f, "Error during epoll call: {}", err),
+            AlreadyExists(pollable) => write!(
+                f,
+                "A handler for the specified pollable {} already exists.",
+                pollable
+            ),
+            NotFound(pollable) => write!(
+                f,
+                "A handler for the specified pollable {} was not found.",
+                pollable
+            ),
+        }
+    }
+}
+
+/// A trait to express the ability to respond to I/O event readiness
+/// using callbacks.
+pub trait Subscriber {
```

Comment:
how about change to "pub trait Subscriber: Send"? so we run multiple IO worker threads to handle epoll events.

Reply:
Could we add some writer/maintainer to the vmm-epoll repo so we could approve the PR?

Reply:
I know this is something that comes up often, but is it very inconvenient if we do not add that restriction to the trait itself, but rather from the outside when using it? Adding the bound here means all subscribers will have to be `Send` which might not fit all potential use cases.

Also, how do you envision running the multiple IO worker threads? Will there be a separate event manager on every thread, or do you prefer some other model?